crates/kornia-pnp/src/refine.rs

```diff
@@ -0,0 +1,333 @@
+        Self {
+            max_iters: DEFAULT_MAX_ITERS,
+            eps: DEFAULT_EPS,
+            lambda_init: DEFAULT_LAMBDA_INIT,
+            lambda_mul: DEFAULT_LAMBDA_MUL,
+            fd_rot_step: FINITE_DIFF_ROT_STEP,
+            fd_trans_base_step: FINITE_DIFF_TRANS_BASE_STEP,
+        }
+    }
+}
+
+/// Refine a pose (rvec, t) with Levenberg–Marquardt to minimize pixel reprojection error.
+///
+/// - `points_world`: World points (N,3)
+/// - `points_image`: Pixel points (N,2)
+/// - `k`: Intrinsics 3x3
+/// - `rvec`: Initial axis-angle rotation (input/output)
+/// - `t`: Initial translation (input/output)
+///
+/// Returns `(rmse, num_iters, converged)` and writes refined `rvec` and `t` in place.
+pub fn refine_pose_lm(
+    points_world: &[[f32; 3]],
+    points_image: &[[f32; 2]],
+    k: &[[f32; 3]; 3],
+    rvec: &mut [f32; 3],
+    t: &mut [f32; 3],
+    params: &LMParams,
+) -> Result<(f32, usize, bool), PnPError> {
+    if points_world.len() != points_image.len() {
+        return Err(PnPError::MismatchedArrayLengths {
+            left_name: "world points",
+            left_len: points_world.len(),
+            right_name: "image points",
+            right_len: points_image.len(),
+        });
+    }
+
+    let n = points_world.len();
+    if n < MIN_CORRESPONDENCES {
+        return Err(PnPError::InsufficientCorrespondences {
+            required: MIN_CORRESPONDENCES,
+            actual: n,
+        });
+    }
+
+    // Parameters vector x = [rx, ry, rz, tx, ty, tz]
+    let mut x = [rvec[0], rvec[1], rvec[2], t[0], t[1], t[2]];
+
+    // Precompute intrinsics vectors for fast projection
+    let fx = k[0][0];
+    let fy = k[1][1];
+    let cx = k[0][2];
+    let cy = k[1][2];
+    let intr_x = Vec3::new(fx, 0.0, cx);
+    let intr_y = Vec3::new(0.0, fy, cy);
+
+    // Projection utility writing residuals in-place to avoid allocations
+    let mut residuals = vec![0.0f32; RESIDUAL_DIM * n];
+    let mut residuals_p = vec![0.0f32; RESIDUAL_DIM * n];
+    let mut residuals_m = vec![0.0f32; RESIDUAL_DIM * n];
+
+    let project_all_in_place = |x: &[f32; 6], out: &mut [f32]| -> f32 {
+        let r_mat = SO3::exp(Vec3A::from_array([x[0], x[1], x[2]])).matrix();
+        let t_vec = Vec3::new(x[3], x[4], x[5]);
+
+        let mut sum_sq = 0.0f32;
+        for (i, (pw_arr, &uv)) in points_world.iter().zip(points_image.iter()).enumerate() {
+            let pw = Vec3::from_array(*pw_arr);
+            let pc = r_mat * pw + t_vec;
+            let inv_z = 1.0 / pc.z;
+            let u_hat = intr_x.dot(pc) * inv_z;
+            let v_hat = intr_y.dot(pc) * inv_z;
+            let du = u_hat - uv[0];
+            let dv = v_hat - uv[1];
+            out[2 * i] = du;
+            out[2 * i + 1] = dv;
+            sum_sq += du.mul_add(du, dv * dv);
+        }
+        sum_sq
+    };
+
+    let mut lambda = params.lambda_init;
+    let mut err_sq_base = project_all_in_place(&x, &mut residuals);
+
+    let mut iters = 0usize;
```

Comment:
```suggestion
    let mut iters = 0;
```
type not recognized by the compiler ?